src/dkg/round3.rs

```diff
@@ -0,0 +1,342 @@
+    identity: &Identity,
+    round2_secret_package: &Round2SecretPackage,
+    round1_public_packages: P,
+    round2_public_packages: Q,
+) -> Result<(KeyPackage, PublicKeyPackage), Error>
+where
+    P: IntoIterator<Item = &'a round1::PublicPackage>,
+    Q: IntoIterator<Item = &'a round2::PublicPackage>,
+{
+    let round1_public_packages = round1_public_packages.into_iter().collect::<Vec<_>>();
+
+    let (min_signers, max_signers) = round2::get_secret_package_signers(round2_secret_package);
+
+    // Ensure that the number of public packages provided matches max_signers
+    if round1_public_packages.len() != max_signers as usize {
+        return Err(Error::InvalidInput(format!(
+            "expected {} public packages, got {}",
+            max_signers,
+            round1_public_packages.len()
+        )));
+    }
+
+    let expected_round1_checksum = round1::input_checksum(
+        min_signers,
+        round1_public_packages.iter().map(|pkg| pkg.identity()),
+    );
+
+    let mut identities = BTreeMap::new();
+    let mut round1_frost_packages: BTreeMap<Identifier, Round1Package> = BTreeMap::new();
+    for public_package in round1_public_packages.clone() {
+        if public_package.checksum() != expected_round1_checksum {
+            return Err(Error::ChecksumError(ChecksumError::DkgPublicPackageError));
+        }
+
+        let identity = public_package.identity();
+        let frost_identifier = identity.to_frost_identifier();
+        let frost_package = public_package.frost_package().clone();
+
+        if round1_frost_packages
+            .insert(frost_identifier, frost_package)
+            .is_some()
+        {
+            return Err(Error::InvalidInput(format!(
+                "multiple public packages provided for identity {}",
+                public_package.identity()
+            )));
+        }
+
+        identities.insert(frost_identifier, identity);
+        round1_frost_packages.insert(
+            public_package.identity().to_frost_identifier(),
+            public_package.frost_package().clone(),
+        );
+    }
+
+    assert_eq!(round1_public_packages.len(), identities.len());
+    assert_eq!(round1_public_packages.len(), round1_frost_packages.len());
+
+    // The public package for `identity` must be excluded from `frost::keys::dkg::part3`
+    // inputs
+    round1_frost_packages
+        .remove(&identity.to_frost_identifier())
+        .expect("missing public package for identity");
+
+    let round2_public_packages = round2_public_packages.into_iter().collect::<Vec<_>>();
+    let expected_round2_checksum =
+        round2::input_checksum(round1_public_packages.iter().map(Borrow::borrow));
+
+    let mut round2_frost_packages: BTreeMap<Identifier, Round2Package> = BTreeMap::new();
+    for public_package in round2_public_packages {
+        if public_package.checksum() != expected_round2_checksum {
+            return Err(Error::ChecksumError(ChecksumError::DkgPublicPackageError));
+        }
+
```

Comment:
do you think it's worthwhile to check here that `self_identity == public_package.recipient_identity`?

Reply:
Added